File: zig/src/indicators/welles_wilder/directional_movement_plus/directional_movement_plus.zig
const std = @import("std");
const math = std.math;


const entities = @import("entities");
const Bar = entities.Bar;
const Quote = entities.Quote;
const Trade = entities.Trade;
const Scalar = entities.Scalar;
const indicator_mod = @import("../../core/indicator.zig");
const build_metadata_mod = @import("../../core/build_metadata.zig");
const identifier_mod = @import("../../core/identifier.zig");
const metadata_mod = @import("../../core/metadata.zig");

const OutputArray = indicator_mod.OutputArray;
const Identifier = identifier_mod.Identifier;
const Metadata = metadata_mod.Metadata;

/// Enumerates the outputs of the Directional Movement Plus indicator.
pub const DirectionalMovementPlusOutput = enum(u8) {
    /// The scalar value of +DM.
    value = 1,
};

/// Parameters for the Directional Movement Plus indicator.
pub const DirectionalMovementPlusParams = struct {
    /// The smoothing length. Must be >= 1. Default is 14.
    length: usize = 14,
};

/// Welles Wilder's Directional Movement Plus (+DM) indicator.
///
/// UpMove = today's high − yesterday's high
/// DownMove = yesterday's low − today's low
/// if UpMove > DownMove and UpMove > 0, then +DM = UpMove, else +DM = 0
///
/// When length > 1, Wilder's smoothing is applied:
///   +DM(n) = previous +DM(n) − previous +DM(n)/n + today's +DM(1)
pub const DirectionalMovementPlus = struct {
    length: usize,
    no_smoothing: bool,
    count: usize,
    previous_high: f64,
    previous_low: f64,
    value: f64,
    accumulator: f64,
    primed: bool,

    pub const Error = error{InvalidLength};

    pub fn init(params: DirectionalMovementPlusParams) Error!DirectionalMovementPlus {
        if (params.length < 1) return error.InvalidLength;

        return .{
            .length = params.length,
            .no_smoothing = params.length == 1,
            .count = 0,
            .previous_high = 0,
            .previous_low = 0,
            .value = math.nan(f64),
            .accumulator = 0,
            .primed = false,
        };
    }

    pub fn deinit(_: *DirectionalMovementPlus) void {}
    pub fn fixSlices(_: *DirectionalMovementPlus) void {}

    /// Core update given high and low values.
    pub fn update(self: *DirectionalMovementPlus, high_in: f64, low_in: f64) f64 {
        if (math.isNan(high_in) or math.isNan(low_in)) {
            return math.nan(f64);
        }

        var high = high_in;
        var low = low_in;
        if (high < low) {
            const tmp = high;
            high = low;
            low = tmp;
        }

        if (self.no_smoothing) {
            if (self.primed) {
                const delta_plus = high - self.previous_high;
                const delta_minus = self.previous_low - low;
                if (delta_plus > 0 and delta_plus > delta_minus) {
                    self.value = delta_plus;
                } else {
                    self.value = 0;
                }
            } else {
                if (self.count > 0) {
                    const delta_plus = high - self.previous_high;
                    const delta_minus = self.previous_low - low;
                    if (delta_plus > 0 and delta_plus > delta_minus) {
                        self.value = delta_plus;
                    } else {
                        self.value = 0;
                    }
                    self.primed = true;
                }
                self.count += 1;
            }
        } else {
            const n: f64 = @floatFromInt(self.length);
            if (self.primed) {
                const delta_plus = high - self.previous_high;
                const delta_minus = self.previous_low - low;
                if (delta_plus > 0 and delta_plus > delta_minus) {
                    self.accumulator += -self.accumulator / n + delta_plus;
                } else {
                    self.accumulator += -self.accumulator / n;
                }
                self.value = self.accumulator;
            } else {
                if (self.count > 0 and self.length >= self.count) {
                    const delta_plus = high - self.previous_high;
                    const delta_minus = self.previous_low - low;
                    if (self.length > self.count) {
                        if (delta_plus > 0 and delta_plus > delta_minus) {
                            self.accumulator += delta_plus;
                        }
                    } else {
                        if (delta_plus > 0 and delta_plus > delta_minus) {
                            self.accumulator += -self.accumulator / n + delta_plus;
                        } else {
                            self.accumulator += -self.accumulator / n;
                        }
                        self.value = self.accumulator;
                        self.primed = true;
                    }
                }
                self.count += 1;
            }
        }

        self.previous_low = low;
        self.previous_high = high;

        return self.value;
    }

    /// Update using a single sample value as substitute for high and low.
    pub fn updateSample(self: *DirectionalMovementPlus, sample: f64) f64 {
        return self.update(sample, sample);
    }

    pub fn isPrimed(self: *const DirectionalMovementPlus) bool {
        return self.primed;
    }

    pub fn getMetadata(_: *const DirectionalMovementPlus, out: *Metadata) void {
        const mnemonic = "+dm";
        const description = "Directional Movement Plus";
        build_metadata_mod.buildMetadata(out, Identifier.directional_movement_plus, mnemonic, description, &.{
            .{ .mnemonic = mnemonic, .description = description },
        });
    }

    fn makeOutput(self: *const DirectionalMovementPlus, time: i64) OutputArray {
        var out = OutputArray{};
        out.append(.{ .scalar = Scalar{ .time = time, .value = self.value } });
        return out;
    }

    pub fn updateScalar(self: *DirectionalMovementPlus, sample: *const Scalar) OutputArray {
        _ = self.update(sample.value, sample.value);
        return self.makeOutput(sample.time);
    }

    pub fn updateBar(self: *DirectionalMovementPlus, sample: *const Bar) OutputArray {
        _ = self.update(sample.high, sample.low);
        return self.makeOutput(sample.time);
    }

    pub fn updateQuote(self: *DirectionalMovementPlus, sample: *const Quote) OutputArray {
        const mid = (sample.bid_price + sample.ask_price) / 2.0;
        _ = self.update(mid, mid);
        return self.makeOutput(sample.time);
    }

    pub fn updateTrade(self: *DirectionalMovementPlus, sample: *const Trade) OutputArray {
        _ = self.update(sample.price, sample.price);
        return self.makeOutput(sample.time);
    }

    // --- Indicator interface ---

    pub fn indicator(self: *DirectionalMovementPlus) indicator_mod.Indicator {
        return indicator_mod.Indicator{
            .ptr = @ptrCast(self),
            .vtable = &.{
                .isPrimed = vtableIsPrimed,
                .metadata = vtableMetadata,
                .updateScalar = vtableUpdateScalar,
                .updateBar = vtableUpdateBar,
                .updateQuote = vtableUpdateQuote,
                .updateTrade = vtableUpdateTrade,
            },
        };
    }

    fn vtableIsPrimed(ptr: *const anyopaque) bool {
        const self: *const DirectionalMovementPlus = @ptrCast(@alignCast(ptr));
        return self.isPrimed();
    }

    fn vtableMetadata(ptr: *const anyopaque, out: *Metadata) void {
        const self: *const DirectionalMovementPlus = @ptrCast(@alignCast(ptr));
        self.getMetadata(out);
    }

    fn vtableUpdateScalar(ptr: *anyopaque, sample: *const Scalar) OutputArray {
        const self: *DirectionalMovementPlus = @ptrCast(@alignCast(ptr));
        return self.updateScalar(sample);
    }

    fn vtableUpdateBar(ptr: *anyopaque, sample: *const Bar) OutputArray {
        const self: *DirectionalMovementPlus = @ptrCast(@alignCast(ptr));
        return self.updateBar(sample);
    }

    fn vtableUpdateQuote(ptr: *anyopaque, sample: *const Quote) OutputArray {
        const self: *DirectionalMovementPlus = @ptrCast(@alignCast(ptr));
        return self.updateQuote(sample);
    }

    fn vtableUpdateTrade(ptr: *anyopaque, sample: *const Trade) OutputArray {
        const self: *DirectionalMovementPlus = @ptrCast(@alignCast(ptr));
        return self.updateTrade(sample);
    }
};

// ============================================================================
// Tests
// ============================================================================

const testing = std.testing;
const testdata = @import("testdata.zig");


fn almostEqual(a: f64, b: f64, tolerance: f64) bool {
    if (math.isNan(a) and math.isNan(b)) return true;
    if (math.isNan(a) or math.isNan(b)) return false;
    return @abs(a - b) <= tolerance;
}

test "DirectionalMovementPlus update length=14" {
    const tolerance = 1e-8;
    var dmp = try DirectionalMovementPlus.init(.{ .length = 14 });

    for (0..testdata.test_high.len) |i| {
        const act = dmp.update(testdata.test_high[i], testdata.test_low[i]);
        const exp = testdata.test_expected_dmp14[i];

        if (math.isNan(exp)) {
            try testing.expect(math.isNan(act));
        } else {
            try testing.expect(!math.isNan(act));
            try testing.expect(almostEqual(act, exp, tolerance));
        }
    }
}

test "DirectionalMovementPlus update length=1" {
    const tolerance = 1e-8;
    var dmp = try DirectionalMovementPlus.init(.{ .length = 1 });

    for (0..testdata.test_high.len) |i| {
        const act = dmp.update(testdata.test_high[i], testdata.test_low[i]);
        const exp = testdata.test_expected_dmp1[i];

        if (math.isNan(exp)) {
            try testing.expect(math.isNan(act));
        } else {
            try testing.expect(!math.isNan(act));
            try testing.expect(almostEqual(act, exp, tolerance));
        }
    }
}

test "DirectionalMovementPlus constructor validation" {
    const result = DirectionalMovementPlus.init(.{ .length = 0 });
    try testing.expectError(error.InvalidLength, result);
}

test "DirectionalMovementPlus isPrimed length=1" {
    var dmp = try DirectionalMovementPlus.init(.{ .length = 1 });

    try testing.expect(!dmp.isPrimed());

    _ = dmp.update(testdata.test_high[0], testdata.test_low[0]);
    try testing.expect(!dmp.isPrimed());

    _ = dmp.update(testdata.test_high[1], testdata.test_low[1]);
    try testing.expect(dmp.isPrimed());
}

test "DirectionalMovementPlus isPrimed length=14" {
    var dmp = try DirectionalMovementPlus.init(.{ .length = 14 });

    for (0..14) |i| {
        _ = dmp.update(testdata.test_high[i], testdata.test_low[i]);
        try testing.expect(!dmp.isPrimed());
    }

    _ = dmp.update(testdata.test_high[14], testdata.test_low[14]);
    try testing.expect(dmp.isPrimed());
}

test "DirectionalMovementPlus NaN passthrough" {
    var dmp = try DirectionalMovementPlus.init(.{ .length = 14 });

    try testing.expect(math.isNan(dmp.update(math.nan(f64), 1)));
    try testing.expect(math.isNan(dmp.update(1, math.nan(f64))));
    try testing.expect(math.isNan(dmp.update(math.nan(f64), math.nan(f64))));
    try testing.expect(math.isNan(dmp.updateSample(math.nan(f64))));
}

test "DirectionalMovementPlus high/low swap" {
    var dmp1 = try DirectionalMovementPlus.init(.{ .length = 1 });
    var dmp2 = try DirectionalMovementPlus.init(.{ .length = 1 });

    _ = dmp1.update(10, 5);
    _ = dmp2.update(5, 10);

    const v1 = dmp1.update(12, 6);
    const v2 = dmp2.update(6, 12);

    try testing.expectEqual(v1, v2);
}

test "DirectionalMovementPlus metadata" {
    var dmp = try DirectionalMovementPlus.init(.{ .length = 14 });
    var meta: Metadata = undefined;
    dmp.getMetadata(&meta);

    try testing.expectEqual(Identifier.directional_movement_plus, meta.identifier);
    try testing.expectEqualStrings("+dm", meta.mnemonic);
    try testing.expectEqualStrings("Directional Movement Plus", meta.description);
    try testing.expectEqual(@as(usize, 1), meta.outputs_len);
}

test "DirectionalMovementPlus updateBar" {
    var dmp = try DirectionalMovementPlus.init(.{ .length = 14 });

    for (0..14) |i| {
        _ = dmp.update(testdata.test_high[i], testdata.test_low[i]);
    }

    const bar = Bar{ .time = 42, .open = 0, .high = testdata.test_high[14], .low = testdata.test_low[14], .close = 0, .volume = 0 };
    const out = dmp.updateBar(&bar);
    try testing.expectEqual(@as(usize, 1), out.len);
}

File: rs/src/indicators/welles_wilder/directional_movement_plus/directional_movement_plus.rs
use crate::entities::bar::Bar;
use crate::entities::quote::Quote;
use crate::entities::scalar::Scalar;
use crate::entities::trade::Trade;
use crate::indicators::core::build_metadata::{build_metadata, OutputText};
use crate::indicators::core::identifier::Identifier;
use crate::indicators::core::indicator::{Indicator, Output};
use crate::indicators::core::metadata::Metadata;

// ---------------------------------------------------------------------------
// Params
// ---------------------------------------------------------------------------

/// Parameters for the Directional Movement Plus indicator.
pub struct DirectionalMovementPlusParams {
    /// Number of time periods. Must be >= 1. Default is 14.
    pub length: usize,
}

impl Default for DirectionalMovementPlusParams {
    fn default() -> Self {
        Self { length: 14 }
    }
}

// ---------------------------------------------------------------------------
// Output
// ---------------------------------------------------------------------------

/// Enumerates the outputs of the DM+ indicator.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum DirectionalMovementPlusOutput {
    /// The scalar value of the directional movement plus.
    Value = 1,
}

// ---------------------------------------------------------------------------
// Indicator
// ---------------------------------------------------------------------------

const DMP_MNEMONIC: &str = "+dm";
const DMP_DESCRIPTION: &str = "Directional Movement Plus";

/// Welles Wilder's Directional Movement Plus indicator.
///
/// +DM measures upward price movement. When the length is greater than 1,
/// Wilder's smoothing method is applied.
pub struct DirectionalMovementPlus {
    length: usize,
    no_smoothing: bool,
    count: usize,
    previous_high: f64,
    previous_low: f64,
    value: f64,
    accumulator: f64,
    primed: bool,
}

impl DirectionalMovementPlus {
    /// Creates a new DirectionalMovementPlus indicator.
    pub fn new(params: &DirectionalMovementPlusParams) -> Result<Self, String> {
        if params.length < 1 {
            return Err(format!("invalid length {}: must be >= 1", params.length));
        }

        Ok(Self {
            length: params.length,
            no_smoothing: params.length == 1,
            count: 0,
            previous_high: 0.0,
            previous_low: 0.0,
            value: f64::NAN,
            accumulator: 0.0,
            primed: false,
        })
    }

    /// Returns the length parameter.
    pub fn length(&self) -> usize {
        self.length
    }

    /// Core update with high and low values.
    pub fn update(&mut self, mut high: f64, mut low: f64) -> f64 {
        if high.is_nan() || low.is_nan() {
            return f64::NAN;
        }

        if high < low {
            std::mem::swap(&mut high, &mut low);
        }

        if self.no_smoothing {
            if self.primed {
                let delta_plus = high - self.previous_high;
                let delta_minus = self.previous_low - low;

                if delta_plus > 0.0 && delta_plus > delta_minus {
                    self.value = delta_plus;
                } else {
                    self.value = 0.0;
                }
            } else {
                if self.count > 0 {
                    let delta_plus = high - self.previous_high;
                    let delta_minus = self.previous_low - low;

                    if delta_plus > 0.0 && delta_plus > delta_minus {
                        self.value = delta_plus;
                    } else {
                        self.value = 0.0;
                    }

                    self.primed = true;
                }

                self.count += 1;
            }
        } else if self.primed {
            let delta_plus = high - self.previous_high;
            let delta_minus = self.previous_low - low;

            if delta_plus > 0.0 && delta_plus > delta_minus {
                self.accumulator += -self.accumulator / self.length as f64 + delta_plus;
            } else {
                self.accumulator += -self.accumulator / self.length as f64;
            }

            self.value = self.accumulator;
        } else {
            if self.count > 0 && self.length >= self.count {
                let delta_plus = high - self.previous_high;
                let delta_minus = self.previous_low - low;

                if self.length > self.count {
                    if delta_plus > 0.0 && delta_plus > delta_minus {
                        self.accumulator += delta_plus;
                    }
                } else {
                    if delta_plus > 0.0 && delta_plus > delta_minus {
                        self.accumulator += -self.accumulator / self.length as f64 + delta_plus;
                    } else {
                        self.accumulator += -self.accumulator / self.length as f64;
                    }

                    self.value = self.accumulator;
                    self.primed = true;
                }
            }

            self.count += 1;
        }

        self.previous_low = low;
        self.previous_high = high;

        self.value
    }

    /// Updates using a single sample value as substitute for high and low.
    pub fn update_sample(&mut self, sample: f64) -> f64 {
        self.update(sample, sample)
    }
}

impl Indicator for DirectionalMovementPlus {
    fn is_primed(&self) -> bool {
        self.primed
    }

    fn metadata(&self) -> Metadata {
        build_metadata(
            Identifier::DirectionalMovementPlus,
            DMP_MNEMONIC,
            DMP_DESCRIPTION,
            &[OutputText {
                mnemonic: DMP_MNEMONIC.to_string(),
                description: DMP_DESCRIPTION.to_string(),
            }],
        )
    }

    fn update_scalar(&mut self, sample: &Scalar) -> Output {
        let v = sample.value;
        let result = self.update(v, v);
        vec![Box::new(Scalar::new(sample.time, result))]
    }

    fn update_bar(&mut self, sample: &Bar) -> Output {
        let result = self.update(sample.high, sample.low);
        vec![Box::new(Scalar::new(sample.time, result))]
    }

    fn update_quote(&mut self, sample: &Quote) -> Output {
        let v = (sample.bid_price + sample.ask_price) / 2.0;
        let result = self.update(v, v);
        vec![Box::new(Scalar::new(sample.time, result))]
    }

    fn update_trade(&mut self, sample: &Trade) -> Output {
        let v = sample.price;
        let result = self.update(v, v);
        vec![Box::new(Scalar::new(sample.time, result))]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::testdata::testdata;
    #[test]
    fn test_dmp_constructor() {
        let dmp = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 14 }).unwrap();
        assert_eq!(dmp.length(), 14);
        assert!(!dmp.is_primed());

        assert!(DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 0 }).is_err());
    }

    #[test]
    fn test_dmp_is_primed() {
        let high = testdata::test_input_high();
        let low = testdata::test_input_low();

        // length=1
        let mut dmp = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 1 }).unwrap();
        assert!(!dmp.is_primed());
        dmp.update(high[0], low[0]);
        assert!(!dmp.is_primed());
        dmp.update(high[1], low[1]);
        assert!(dmp.is_primed());

        // length=14
        let mut dmp = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 14 }).unwrap();
        for i in 0..14 {
            dmp.update(high[i], low[i]);
            assert!(!dmp.is_primed(), "[{}] should not be primed yet", i);
        }
        dmp.update(high[14], low[14]);
        assert!(dmp.is_primed());
    }

    #[test]
    fn test_dmp_update() {
        let tolerance = 1e-8;
        let high = testdata::test_input_high();
        let low = testdata::test_input_low();
        let expected = testdata::test_expected_dmp14();
        let mut dmp = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 14 }).unwrap();

        for i in 0..high.len() {
            let act = dmp.update(high[i], low[i]);

            if expected[i].is_nan() {
                assert!(act.is_nan(), "[{}] expected NaN, got {}", i, act);
                continue;
            }

            assert!(!act.is_nan(), "[{}] expected {}, got NaN", i, expected[i]);
            assert!(
                (act - expected[i]).abs() < tolerance,
                "[{}] expected {}, got {}",
                i, expected[i], act
            );
        }
    }

    #[test]
    fn test_dmp_length_1() {
        let tolerance = 1e-8;
        let high = testdata::test_input_high();
        let low = testdata::test_input_low();
        let expected = testdata::test_expected_dmp1();
        let mut dmp = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 1 }).unwrap();

        for i in 0..high.len() {
            let act = dmp.update(high[i], low[i]);

            if expected[i].is_nan() {
                assert!(act.is_nan(), "[{}] expected NaN, got {}", i, act);
                continue;
            }

            assert!(!act.is_nan(), "[{}] expected {}, got NaN", i, expected[i]);
            assert!(
                (act - expected[i]).abs() < tolerance,
                "[{}] expected {}, got {}",
                i, expected[i], act
            );
        }
    }

    #[test]
    fn test_dmp_nan_passthrough() {
        let mut dmp = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 14 }).unwrap();
        assert!(dmp.update(f64::NAN, 1.0).is_nan());
        assert!(dmp.update(1.0, f64::NAN).is_nan());
        assert!(dmp.update(f64::NAN, f64::NAN).is_nan());
        assert!(dmp.update_sample(f64::NAN).is_nan());
    }

    #[test]
    fn test_dmp_high_low_swap() {
        let mut dmp1 = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 1 }).unwrap();
        let mut dmp2 = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 1 }).unwrap();

        dmp1.update(10.0, 5.0);
        dmp2.update(5.0, 10.0);

        let v1 = dmp1.update(12.0, 6.0);
        let v2 = dmp2.update(6.0, 12.0);

        assert_eq!(v1, v2);
    }

    #[test]
    fn test_dmp_zero_inputs() {
        let mut dmp = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 10 }).unwrap();

        for _ in 0..20 {
            dmp.update_sample(0.0);
        }

        assert!(dmp.is_primed());
    }

    #[test]
    fn test_dmp_metadata() {
        let dmp = DirectionalMovementPlus::new(&DirectionalMovementPlusParams { length: 14 }).unwrap();
        let meta = dmp.metadata();
        assert_eq!(meta.identifier, Identifier::DirectionalMovementPlus);
        assert_eq!(meta.mnemonic, "+dm");
        assert_eq!(meta.description, "Directional Movement Plus");
        assert_eq!(meta.outputs.len(), 1);
        assert_eq!(meta.outputs[0].mnemonic, "+dm");
        assert_eq!(meta.outputs[0].description, "Directional Movement Plus");
    }
}
